p-ata/src/processor.rs

```diff
@@ -0,0 +1,974 @@
+    #[cfg(target_os = "solana")]
+    {
+        const ED25519_CURVE_ID: u64 = 0;
+
+        let point_addr = address.as_ref().as_ptr();
+
+        // SAFETY: We're passing valid pointers to the syscall
+        // The syscall directly returns the validation result:
+        // - 0 means point is ON the curve (valid)
+        // - 1 means point is OFF the curve (invalid)
+        // - any other value means error
+        let syscall_result = unsafe {
+            sol_curve_validate_point(ED25519_CURVE_ID, point_addr, core::ptr::null_mut())
+        };
+
+        syscall_result == 1
+    }
+    #[cfg(all(not(target_os = "solana"), test))]
+    {
+        // Host build (tests, benches): replicate the on-chain `sol_curve_validate_point` logic
+        // using curve25519-dalek. A pubkey is "off-curve" if it cannot be decompressed into
+        // an Edwards point.
+
+        curve25519_dalek::edwards::CompressedEdwardsY(*address)
+            .decompress()
+            .is_none()
+    }
+    #[cfg(all(not(target_os = "solana"), not(test)))]
+    {
+        false
+    }
+}
+
+/// Validate an expected bump and ensure no better canonical bump exists.
+///
+/// Given an expected bump, this function verifies that no higher bump value produces
+/// a valid (off-curve) PDA address. This prevents creation of non-canonical ATAs
+/// by rejecting sub-optimal bump seeds.
+///
+/// ## Arguments
+///
+/// * `seeds` - The PDA derivation seeds (wallet, token_program, mint)
+/// * `program_id` - The program ID for PDA derivation  
+/// * `expected_bump` - The bump value provided by the caller
+///
+/// ## Returns
+///
+/// * `None` - No better bump exists, the expected_bump is canonical
+/// * `Some((address, bump))` - A better bump was found, returns the canonical address and bump
+///
+/// ## Security
+///
+/// This function is critical for preventing PDA canonicality attacks. It ensures
+/// that only the highest valid bump can be used, maintaining deterministic
+/// address derivation across all clients.
+#[inline(always)]
+pub(crate) fn reject_if_better_valid_bump_exists(
+    seeds: &[&[u8]; 3],
+    program_id: &Pubkey,
+    expected_bump: u8,
+) -> Result<(), ProgramError> {
+    // Optimization: Only verify no better bump exists. Don't require expected_bump to
+    // yield an off-curve address. This saves significant compute units while still
+    // preventing non-canonical addresses.
+    //
+    // Caller must ensure that `expected_bump` is off-curve, either via downstream failure
+    // (i.e. syscalls that will fail) or by calling `is_off_curve`.
+    let mut better_bump = 255;
+    while better_bump > expected_bump {
+        let maybe_better_address = derive_address::<3>(seeds, Some(better_bump), program_id);
+        if is_off_curve(&maybe_better_address) {
+            log!("Canonical address does not match provided address. Canonical bump is {}, with address {}.", better_bump, &maybe_better_address);
```

Comment:
Not sure if you are expecting the `log!` to print the pubkey in base58 – as it is it will print the bytes.